### PR TITLE
CC-12:  Input controls

### PR DESCRIPTION
**Pull request by ankitp-atharva**
https://github.com/Components-care/backlog/issues/12
New controls added for:
- NumberInteger
- NumberDecimal
- DateInput
- DateTimeInput
- CurrencyInput

### CODE REVIEW

**Review by TQjona — CHANGES_REQUESTED** [review]
Lacking value formatting of decimals and integer inputs using globalize.js. Furthermore these inputs need custom onChange events which have number values instead of string values. The inputs also should be controllable via props (which isn't the case for the integer input). Date/Time inputs also lack translation and automatic localization, create new localized components for them (like the LocalizedDateTimePicker component).

**Comment by ankitp-atharva** [comment]
> Lacking value formatting of decimals and integer inputs using globalize.js. Furthermore these inputs need custom onChange events which have number values instead of string values. The inputs also should be controllable via props (which isn't the case for the integer input). Date/Time inputs also lack translation and automatic localization, create new localized components for them (like the LocalizedDateTimePicker component).

@TQjona ,  Added formatting, onChange props and required changes as suggested. Please review.

File: src/standalone/UIKit/InputControls/CurrencyInput.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import { TextFieldProps } from "@material-ui/core";
+import TextFieldWithHelp from "../TextFieldWithHelp";
+import { UIInputProps } from "../CommonStyles";
+import { getGlobalized } from "../../../globalize";
+
+export interface CurrencyInputProps extends UIInputProps {
+	/**
+	 * The text for info icon
+	 */
+	infoText?: React.ReactNode;
```
Comment (TQjona) [inline src/standalone/UIKit/InputControls/CurrencyInput.tsx:11]: why is `infoText` and `important` specified on these property type defs? just `extends TextFieldWithHelpProps`

File: src/standalone/UIKit/InputControls/CurrencyInput.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	/**
+	 * Boolean flag to make css important
+	 */
+	important?: boolean;
+	/**
+	 * The currency to be used in formatting
+	 */
+	currency: string;
+}
```
Comment (TQjona) [inline src/standalone/UIKit/InputControls/CurrencyInput.tsx:20]: missing (to be added to **all** input elements): 
```ts
value?: T
onChange?: (newValue: T) => void`;
onBlur?: React.FocusEventHandler;
``` 
T = data type of input => 
- Currency, Integer, Decimal => number | null
- Text => string
- Date, DateTime => MaterialUiPickersDate

If value != undefined the input is a controlled input => doesn't use internal state
If value == undefined the input is uncontrolled => uses internal state to keep track of it's value. internal state is of type T, formatting happens by globalize.js before passing the raw value to the TextField component

File: src/standalone/UIKit/InputControls/CurrencyInput.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+const CurrencyInput = (props: CurrencyInputProps & TextFieldProps) => {
+	const { currency, infoText, important, ...muiProps } = props;
+	const [number, setNumber] = useState("");
+	const onChange = async (event: React.ChangeEvent<HTMLInputElement>) => {
+		let value = event.target.value;
+		if (value != "") {
+			if (value.split(".").length > 2) {
+				value = number;
+			}
+			const numericValue = Number(value.replace(/[^0-9.]/g, ""));
```
Comment (TQjona) [inline src/standalone/UIKit/InputControls/CurrencyInput.tsx:31]: use the globalize.js currency/decimal/etc parser

File: src/standalone/UIKit/InputControls/PhoneNumberInput.tsx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	const [error, setError] = useState(false);
+
+	const updateValue = useCallback(
+		(event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
+			let valid = true;
+			let phoneNumber = event.target.value.replace(/[^0-9]/g, "");
+			const numericValue = Number(phoneNumber);
+			const newPhoneNumber = parsePhoneNumberFromString(
+				phoneNumber,
+				countryCode || "DE"
```
Comment (TQjona) [inline src/standalone/UIKit/InputControls/PhoneNumberInput.tsx:49]: why default to `DE`? make it a required prop instead

File: src/standalone/UIKit/InputControls/PhoneNumberInput.tsx [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+			);
+			if (newPhoneNumber) {
+				valid = newPhoneNumber.isValid();
+				if (valid) phoneNumber = newPhoneNumber.number as string;
+			}
+			setError(!valid);
+			setValue(phoneNumber);
+			getValue(numericValue);
```
Comment (TQjona) [inline src/standalone/UIKit/InputControls/PhoneNumberInput.tsx:57]: remove the numericValue, international phone numbers cannot be represented using numbers (ex: +49 123 123456 => 0049123123456 => trailing zeros are removed => invalid phone number)

File: src/standalone/UIKit/InputControls/PhoneNumberInput.tsx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+		},
+		[countryCode, getValue, setValue]
+	);
+	const handleChange = useCallback(
+		(event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
```
Comment (TQjona) [inline src/standalone/UIKit/InputControls/PhoneNumberInput.tsx:62]: this function does nothing? replace it with a direct call to `updateValue`

File: src/standalone/UIKit/InputControls/PhoneNumberInput.tsx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+		<div>
+			<TextFieldWithHelp
+				{...muiProps}
+				autoFocus
+				important={important}
+				value={value}
+				onFocus={updateValue}
+				onChange={handleChange}
+				error={!!error}
+				helperText={error && "Enter correct phone number"}
```
Comment (TQjona) [inline src/standalone/UIKit/InputControls/PhoneNumberInput.tsx:78]: error msg not localized

File: src/standalone/UIKit/TextFieldWithHelp.tsx [hunk 8]
```diff
@@ -5,10 +5,11 @@ import { InputLabelConfig, UIInputProps, useInputStyles } from "./CommonStyles";
 
 export interface TextFieldWithHelpProps extends UIInputProps {
 	infoText?: React.ReactNode;
+	type?: string;
```
Comment (TQjona) [inline src/standalone/UIKit/TextFieldWithHelp.tsx:8]: pretty sure this is not needed as it's supplied by TextFieldProps

**Review by TQjona — CHANGES_REQUESTED** [review]

File: src/standalone/UIKit/InputControls/CurrencyInput.tsx (hunk 2 above)
Comment (ankitp-atharva) [reply src/standalone/UIKit/InputControls/CurrencyInput.tsx:17]: @TQjona ,  Currency, Integer, Decimal => number | null
Here we need string type, as we are going to display formatted value in text input. We are returning number in getValue() callback.

Comment (TQjona) [reply src/standalone/UIKit/InputControls/CurrencyInput.tsx:17]: @ankitp-atharva How about:
```tsx
<TextField
  value={formatNumber(value)}
  onChange={parseNumber}
/>
```

File: src/standalone/LocalizedDateTimePickers/LocalizedDateTimePicker.tsx [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+		invalidDateMessage={ccI18n.t(
+			"backend-integration.model.types.renderers.date.labels.invalid-date"
+		)}
+		minDateMessage={ccI18n.t(
+			"backend-integration.model.types.renderers.date.labels.min-date"
+		)}
+		maxDateMessage={ccI18n.t(
+			"backend-integration.model.types.renderers.date.labels.max-date"
+		)}
+		format={"L hh:mm A"}
```
Comment (TQjona) [inline src/standalone/LocalizedDateTimePickers/LocalizedDateTimePicker.tsx:46]: format not localized

File: src/standalone/UIKit/InputControls/NumberDecimal.tsx [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+	onChange?: (newValue: T) => void;
+	onBlur?: React.FocusEventHandler;
+}
+
+const NumberDecimal = (
+	props: NumberDecimalProps<string> & TextFieldWithHelpProps & TextFieldProps
+) => {
+	const { getValue, setValue, infoText, important, ...muiProps } = props;
+
+	const parseNumber = (s: string) => {
```
Comment (TQjona) [inline src/standalone/UIKit/InputControls/NumberDecimal.tsx:32]: also use globalize.js number parser here

**Review by TQjona — CHANGES_REQUESTED** [review]

File: src/standalone/UIKit/InputControls/CurrencyInput.tsx (hunk 2 above)
Comment (ankitp-atharva) [reply src/standalone/UIKit/InputControls/CurrencyInput.tsx:17]: Updated as suggested

File: src/utils/getDecimalSeparator.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import ccI18n from "../i18n";
+
+const getDecimalSeparator = (): string => {
```
Comment (TQjona) [inline src/utils/getDecimalSeparator.ts:3]: `getSeperator("decimal")` does the same as this function

File: src/utils/getSeparator.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import ccI18n from "../i18n";
+
+/**
+ * Gets the group seperator from number by language
+ * @param seperatorType The seperator type (eg: group)
+ */
+const getSeparator = (separatorType: string): string => {
```
Comment (TQjona) [inline src/utils/getSeparator.ts:7]: rename to `getNumberSeparator`

File: src/stories/standalone/UIKit/InputControls/PhoneNumberInput.story.tsx [hunk 13]
```diff
@@ -0,0 +1,37 @@
+import React, { useState } from "react";
+import { boolean, text, select } from "@storybook/addon-knobs";
+import PhoneNumberInput from "../../../../standalone/UIKit/InputControls/PhoneNumberInput";
+import { CountryCode } from "libphonenumber-js";
+import { action } from "@storybook/addon-actions";
+
+export const PhoneNumberInputStory = (): React.ReactElement => {
+	const countryCodes = ["US", "DE", "FR", "RU"] as CountryCode[];
```
Comment (TQjona) [inline src/stories/standalone/UIKit/InputControls/PhoneNumberInput.story.tsx:8]: don't cast here, just manually define the type of `countryCodes` => `const countryCodes: CountryCode[] = [...]`

File: src/stories/standalone/UIKit/InputControls/DateTimeInput.story.tsx [hunk 14]
```diff
@@ -0,0 +1,45 @@
+import React, { useState } from "react";
+import { boolean, text } from "@storybook/addon-knobs";
+import moment, { Moment } from "moment";
+import { action } from "@storybook/addon-actions";
+import DateTimeInput from "../../../../standalone/UIKit/InputControls/DateTimeInput";
+import { MaterialUiPickersDate } from "@material-ui/pickers/typings/date";
+
+export const DateTimeInputStory = (): React.ReactElement => {
+	const [selectedDate, setSelectedDate] = useState(moment());
```
Comment (TQjona) [inline src/stories/standalone/UIKit/InputControls/DateTimeInput.story.tsx:9]: why are you using moment here? This is a normal Date object or null. `useState<MaterialUiPickersDate>(null)`

File: src/stories/standalone/UIKit/InputControls/DateInput.story.tsx [hunk 15]
```diff
@@ -0,0 +1,44 @@
+import React, { useState } from "react";
+import { boolean, text } from "@storybook/addon-knobs";
+import moment, { Moment } from "moment";
+import { action } from "@storybook/addon-actions";
+import DateInput from "../../../../standalone/UIKit/InputControls/DateInput";
+import { MaterialUiPickersDate } from "@material-ui/pickers/typings/date";
+
+export const DateInputStory = (): React.ReactElement => {
+	const [selectedDate, setSelectedDate] = useState(moment());
```
Comment (TQjona) [inline src/stories/standalone/UIKit/InputControls/DateInput.story.tsx:9]: see DateTimeInputStory

File: src/standalone/UIKit/InputControls/PhoneNumberInput.tsx [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+export interface PhoneNumberInputProps extends TextFieldWithHelpProps {
+	/**
+	 * The CountryCode for formatting phone number
+	 */
+	countryCode: CountryCode;
+	/**
+	 * The current value or null if not set
+	 */
+	value: number | null;
```
Comment (TQjona) [inline src/standalone/UIKit/InputControls/PhoneNumberInput.tsx:17]: You cannot represent a phone number as number, it has to be a string

File: src/standalone/UIKit/InputControls/NumberDecimal.tsx [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+			ccI18n.off("languageChanged", updateGlobalized);
+		};
+	}, []);
+
+	// on change handling
+	const handleChange = useCallback(
+		(event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
+			if (!onChange) return;
+
+			const num = event.target.value;
```
Comment (TQjona) [inline src/standalone/UIKit/InputControls/NumberDecimal.tsx:57]:
```tsx
const num = event.target.value
    .replace(new RegExp("[^0-9" + getSeparator("decimal") + "]", "g"), "")
    .replace(new RegExp(getSeparator("decimal"), "g"), ".");
if (num !== "") {
  onChange(parseFloat(num));
} else {
  onChange(null);
}

File: src/standalone/UIKit/InputControls/NumberDecimal.tsx [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+import TextFieldWithHelp, {
+	TextFieldWithHelpProps,
+} from "../TextFieldWithHelp";
+import { getGlobalized } from "../../../globalize";
+import Globalize from "globalize/dist/globalize";
+import ccI18n from "../../../i18n";
+import { getSeparator, getDecimalSeparator } from "../../../utils";
+
+export interface NumberDecimalProps extends TextFieldWithHelpProps {
+	/**
```
Comment (TQjona) [inline src/standalone/UIKit/InputControls/NumberDecimal.tsx:12]: remove this comment

File: src/standalone/UIKit/InputControls/CurrencyInput.tsx [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+			ccI18n.off("languageChanged", updateGlobalized);
+		};
+	}, []);
+
+	// on change handling
+	const handleChange = useCallback(
+		(event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
+			if (!onChange) return;
+
+			const num = event.target.value;
```
Comment (TQjona) [inline src/standalone/UIKit/InputControls/CurrencyInput.tsx:58]: see Decimal Input

**Review by TQjona — CHANGES_REQUESTED** [review]

File: src/standalone/UIKit/InputControls/NumberDecimal.tsx (hunk 18 above)
Comment (ankitp-atharva) [reply src/standalone/UIKit/InputControls/NumberDecimal.tsx:12]: Updated

File: src/standalone/UIKit/InputControls/NumberDecimal.tsx (hunk 17 above)
Comment (ankitp-atharva) [reply src/standalone/UIKit/InputControls/NumberDecimal.tsx]: Updated as above logic

File: src/standalone/UIKit/InputControls/CurrencyInput.tsx (hunk 19 above)
Comment (ankitp-atharva) [reply src/standalone/UIKit/InputControls/CurrencyInput.tsx]: Updated

File: src/stories/standalone/UIKit/InputControls/DateTimeInput.story.tsx (hunk 14 above)
Comment (ankitp-atharva) [reply src/stories/standalone/UIKit/InputControls/DateTimeInput.story.tsx]: Updated

File: src/utils/getDecimalSeparator.ts (hunk 11 above)
Comment (TQjona) [reply src/utils/getDecimalSeparator.ts:3]: means replace all occurances of this function with the above mentioned and remove it here

Comment (ankitp-atharva) [reply src/utils/getDecimalSeparator.ts]: @TQjona  - Removed this file

**Review by TQjona — APPROVED** [review]

**Pull request merged**